### ISSUE TITLE
[VoodooI2CHID] Fixing Physical Buttons on Trackpad

### ISSUE THREAD

**Issue by UsedDiscord** [Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh#8]
So, after updating Kexts... the trackpad works!

CONNEXION : #4 #3

**Comment by UsedDiscord** [comment]
#4 #3

**Comment by Qonfused** [comment]
Does it work with gestures or just as a ps2 mouse (right/left click only)?

It does make sense that updating OpenCore + Kexts though should fix some instability. OpenCore versions 0.7.8 to 0.8.2 should be mostly the same; if you want to boot Ventura, you'll also need OpenCore 0.8.3 and Lilu 1.6.2.

**Comment by UsedDiscord** [comment]
So... some explanations. I updated, and deleted some Kexts that are completely useless... but also temporarily disable some kexts of VoodooI2C

So : 

### **UPDATED** :

- AirportItlwm.kext -> https://github.com/OpenIntelWireless/itlwm/releases/download/v2.1.0/AirportItlwm_v2.1.0_stable_Monterey.kext.zip

- AppleALC.kext -> https://github.com/acidanthera/AppleALC/releases/download/1.7.4/AppleALC-1.7.4-RELEASE.zip

- BlueToolFixup.kext -> https://github.com/acidanthera/BrcmPatchRAM/releases/download/2.6.3/BrcmPatchRAM-2.6.3-RELEASE.zip

- BrightnessKeys.kext -> https://github.com/acidanthera/BrightnessKeys/releases/download/1.0.2/BrightnessKeys-1.0.2-RELEASE.zip

- CodecCommander.kext (il me semble en erreur lors du démarrage) -> https://bitbucket.org/RehabMan/os-x-eapd-codec-commander/downloads/RehabMan-CodecCommander-2018-1003.zip

- CPUFriend.kext -> https://github.com/acidanthera/CPUFriend/releases/download/1.2.6/CPUFriend-1.2.6-RELEASE.zip
 
- CpuTscSync.kext -> https://github.com/acidanthera/CpuTscSync/releases/download/1.0.9/CpuTscSync-1.0.9-RELEASE.zip

- DebugEnhacer.kext -> https://github.com/acidanthera/DebugEnhancer/releases/download/1.0.7/DebugEnhancer-1.0.7-RELEASE.zip

- ECEnabler.kext -> https://github.com/1Revenger1/ECEnabler/releases/download/1.0.3/ECEnabler-1.0.3-RELEASE.zip

- FeatureUnlock.kext -> https://github.com/acidanthera/FeatureUnlock/releases/download/1.0.9/FeatureUnlock-1.0.9-RELEASE.zip

- HibernationFixup.kext -> https://github.com/acidanthera/HibernationFixup/releases/download/1.4.6/HibernationFixup-1.4.6-RELEASE.zip

- IntelBluetoothFirmware.kext -> https://github.com/OpenIntelWireless/IntelBluetoothFirmware/releases/download/v2.2.0/IntelBluetooth-v2.2.0.zip

- IntelBTPPatcher.kext (NEW !!) -> https://github.com/OpenIntelWireless/IntelBluetoothFirmware/releases/download/v2.2.0/IntelBluetooth-v2.2.0.zip

- Lilu.kext (DEBUG VERSION!) -> https://github.com/acidanthera/Lilu/releases/download/1.6.2/Lilu-1.6.2-DEBUG.zip

- NoTouchID.kext -> https://github.com/al3xtjames/NoTouchID/releases/download/1.0.3/NoTouchID-1.0.3-RELEASE.zip

- NVMeFix.kext -> https://github.com/acidanthera/NVMeFix/releases/download/1.1.0/NVMeFix-1.1.0-RELEASE.zip

- RestrictEvents.kext -> https://github.com/acidanthera/RestrictEvents/releases/download/1.0.8/RestrictEvents-1.0.8-RELEASE.zip

- RTCMemoryFixup.kext -> https://github.com/acidanthera/RTCMemoryFixup/releases/download/1.0.7/RTCMemoryFixup-1.0.7-RELEASE.zip

- SMC*.kext (include : BatteryManager, Processor, LightSenor & SuperIO) (Break?) : -> https://github.com/acidanthera/VirtualSMC/releases/download/1.3.0/VirtualSMC-1.3.0-RELEASE.zip

- VirtualSMC -> https://github.com/acidanthera/VirtualSMC/releases/download/1.3.0/VirtualSMC-1.3.0-RELEASE.zip

- VoodooI2C.kext & VoodooI2CHID.kext -> https://github.com/VoodooI2C/VoodooI2C/releases/download/2.7/VoodooI2C-2.7.zip

- VoodooPS2Controller.kext -> https://github.com/acidanthera/VoodooPS2/releases/download/2.2.9/VoodooPS2Controller-2.2.9-RELEASE.zip

- WhateverGreen.kext (DEBUG Version) -> https://github.com/acidanthera/WhateverGreen/releases/download/1.6.1/WhateverGreen-1.6.1-DEBUG.zip

- XHCI-Unsupported.kext -> https://github.com/johnlimabravo/XHCI-unsupported/blob/master/XHCI-unsupported.kext.zip

### **NOT UPDATED** :

- ACPIDebug.kext (deactivate in config.plist)
- AnyiSightCam.kext
- AsusSMC.kext
- CPUFriendDataProvider.kext
- DiskArbitrationFixup.kext
- HoRNDIS.kext
- IOElectrify.kext
- USBInjectAll.kext
- USBPorts.kext

### **REMOVED** : 

- Airportlwm-BigSur.kext
- Airportlwm-Catalina.Kext
- IntelBluetoothInjector.kext
- RtWlanU.kext
- RtWlanU1827.kext

### **Temporarily suspended (Temporarily removed)** : 

- VoodooI2CHID-AsTouchScreen.kext
- VoodooI2CHID-AsTrackpad.kext

**Comment by UsedDiscord** [comment]
About the trackpad: 

### **Gesture**: 

3 fingers: open the window manager & search with Siri
2 fingers: up & down 

### **Does not work**: 

2 fingers: secondary click

### **Touch**: 

Approximately OK... but I have the impression that if we touch the touch, we will have to use the touch forever until the reboot of the PC..... we would be potentially obliged to reboot the pc to get back the hand on the trackpad...

**Comment by UsedDiscord** [comment]
> #4 #3
> 
> I write a text and i upload it

Thanks to GitHub for not letting me upload my zip file for 0,3Mb too much haha...

[Kexts.zip](https://drive.google.com/file/d/1BfhrRicqq627LMGVOYHXHL20Gib7u18a/view?usp=sharing)
[config.plist](https://drive.google.com/file/d/1KJvNkVT--atU03qQeg42fTXL6btCVVsO/view?usp=sharing)

**Don't forget to change your series values please!**

As @Qonfused said :
> Warning Please follow the below instructions before using this repository.

> Please download [GenSMBIOS] (https://github.com/corpnewt/GenSMBIOS)  and follow [this guide](https://dortania.github.io/OpenCore-Post-Install/universal/iservices.html#using-gensmbios) for generating SMBIOS data for enabling iServices functionality.


**For the screenpad :** 
but no progress concerning the ScreenPad for the moment...

**Comment by UsedDiscord** [comment]
> It does make sense that updating OpenCore + Kexts though should fix some instability. OpenCore versions 0.7.8 to 0.8.2 should be mostly the same; if you want to boot Ventura, you'll also need OpenCore 0.8.3 and Lilu 1.6.2.

So I already managed to get started on Monterey thanks to you, I'm not going to take a chance by screwing up in Ventura heh...

**Comment by UsedDiscord** [comment]
I made a mistake in the configuration, it is replaced.

If you have an error when starting MacOS, just download "config.plist" again ! and replace it ! I forgot a letter, sorry about the problem... I was tired haha...

ERROR : (line 1500)
> <string>Contents/MacOS/IntelBTPacher</string>

FIXING BY : 

> <string>Contents/MacOS/IntelBTPatcher</string>

New config.plist uploaded !

[config.plist](https://drive.google.com/file/d/1KJvNkVT--atU03qQeg42fTXL6btCVVsO/view?usp=sharing)

**Comment by Qonfused** [comment]
So far I've updated these kexts in the [`#update-opencore-config`](https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/tree/update-opencore-config) branch. Some things to note is that IOElectrify.kext and USBInjectAll.kext aren't needed (they're already disabled in your config though).

**Comment by UsedDiscord** [comment]
Oh I hadn't noticed, thank you!Le 26 août 2022 23:37, Cory Bennett ***@***.***> a écrit :
So far I've updated these kexts in the #update-opencore-config branch. Some things to note is that IOElectrify.kext and USBInjectAll.kext aren't needed (they're already disabled in your config though).

**Comment by Qonfused** [comment]
I wasn't able to reproduce the same behavior after updating all the kexts, though I may need to re-test that after I reverted that commit by mistake (I don't recall whether I actually updated the Voodoo kexts at that point, though I recall testing without updating them).

I did note that VoodooI2C and VoodooI2CHID are both used for the trackpad and touchscreen, and VoodooPS2 is used only for the keyboard. I didn't see the `VoodooI2CHID-AsTouchscreen.kext` or the `VoodooI2CHID-AsTouchscreen.kext` loaded in my config (or the original config). I'm not sure what their purpose is; they're likely artifacts from previous testing.

I'm assuming that loading the root kext should be done before loading any of it's plugins, so there may be an improper loading order in my config. Below is a table of their index (only the order matters) and some relevant entry properties, shortened for brevity:

Idx | Source Kext | Executable | Kext Plugin | Enabled
--- | --- | --- | --- | ---
34 | VoodooI2C.kext | `Contents/MacOS/VoodooGPIO` | VoodooGPIO.kext | Yes
35 | VoodooI2C.kext | `Contents/MacOS/VoodooI2CServices` | VoodooI2CServices | Yes
36 | VoodooI2C.kext | `Contents/MacOS/VoodooInput` | VoodooInput.kext | Yes
37 | VoodooI2C.kext | `Contents/MacOS/VoodooI2C` | | Yes
38 | VoodooI2CHID.kext | `Contents/MacOS/VoodooI2CHID` | | Yes
39 | VoodooPS2Controller | `Contents/MacOS/VoodooPS2Controller` | | Yes
40 | VoodooPS2Controller | `Contents/Plugins/VoodooPS2Keyboard` | VoodooPS2Keyboard.kext | Yes
41 | VoodooPS2Controller | `Contents/Plugins/VoodooPS2Mouse` | VoodooPS2Mouse.kext | `No`
42 | VoodooPS2Controller | `Contents/Plugins/VoodooPS2Trackpad` | VoodooPS2Trackpad.kext | `No`
43 | VoodooPS2Controller | `Contents/Plugins/VoodooInput` | VoodooPS2Input.kext | `No`

I figure this is caused by some fixes for power-optimization (e.g. GPIO pinning for the trackpad and touchscreen). In #7 I'll try to figure out which of those ACPI patches are might be interfering or are simply incompatible. I do notice an error mentioning `AddressingMode7Bit` and `I2C1` or something along those lines, which are indicative of some kind of GPIO pinning recognizable from [the voodoo guide](https://github.com/VoodooI2C/VoodooI2C/blob/master/Documentation/GPIO%20Pinning.md).

**Comment by UsedDiscord** [comment]
Missclick*

**Comment by Qonfused** [comment]
Huh, seems there is a guide for fixing the GPI0 device in the ACPI patching section of the guide [here](https://dortania.github.io/Getting-Started-With-ACPI/Laptops/trackpad-methods/manual.html#checking-gpi0). If applicable, I'll diff my changes from what may already be implemented, but I'd have to verify that the device is working correctly before any additional GPI0-related ACPI patches can get loaded.

Will have to wait until tomorrow to have time to work on that, but it is a much easier process than I imagined (and hopefully that's the issue if the device works fine when it works).

**Comment by Qonfused** [comment]
Noticing in #10 that the left click button maps to right click, and the right click button maps to middle click.

**Comment by Qonfused** [comment]
There is some conflict that prevents touchscreen from working several seconds after startup, even with the `-vi2c-force-polling` boot arg. Not a new issue, but one I didn't expect to persist.

**Comment by UsedDiscord** [comment]
> There is some conflict that prevents touchscreen from working several seconds after startup,

A few seconds is fine... Worse would have been if it had been blocked for a long time...

**Comment by Qonfused** [comment]
Sorry I meant *after several seconds; it can occasionally happen to the trackpad I've noticed, but both are resolved after closing the lid for a few seconds.

**Comment by Qonfused** [comment]
Moved touchscreen issue to #12.

**Comment by UsedDiscord** [comment]
So, if we use Karabiner as planned, and we use 2 simple modifiers :

A) Button 2 -> Button 3 
B) Button 3 -> Button 2

![Capture d’écran 2022-11-30 à 16 10 30](https://user-images.githubusercontent.com/57113628/204834874-9640eeb7-ebad-4d6e-850a-8f5249b238c6.png)

(As like on my screen)

this will solve the problem for the moment, but I have to test if it's normal that my trackpad goes very fast and not as usual...

**Comment by UsedDiscord** [comment]
update : Doesn't work right now... i need more time...

**Comment by UsedDiscord** [comment]
In looking on Issues about Karabiner, i'have found the SAME problem about we, [Karabiner-Element Issues 2370](https://github.com/pqrs-org/Karabiner-Elements/issues/2370)

So, we need just to edit karabiner.json with that :

```json
"simple_modifications": [
  {
    "from": {
      "pointing_button": "button2"
    },
    "to": {
      "pointing_button": "button1"
    }
  },
  {
    "from": {
      "pointing_button": "button1"
    },
    "to": {
      "pointing_button": "button2"
    }
  }
]
```

_Originally posted by @kirintwn in https://github.com/pqrs-org/Karabiner-Elements/issues/2370#issuecomment-674669989_

**Comment by Qonfused** [comment]
Gonna look into disabling force touch by default as well:
https://discord.com/channels/186648463541272576/478720084072988672/1052709370259054683

**Comment by Qonfused** [comment]
Managed to fix the buttons with a reorder of kext plugins and adding voodoops2trackpad + voodooinput in https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/commit/0b6711556207d49135466101fb8e217ab68704dd. I downgraded voodoo kext versions as part of this, but the same fix as proposed above should be applied when updating to newer versions.

**Comment by Qonfused** [comment]
PR merge on Jul 22 seems to address this issue: https://github.com/VoodooI2C/VoodooI2C/commit/9ab98319b6411c4a4822cc12f840df85cc684bfc
> **Update VoodooI2C Satellites and support for physical buttons** (https://github.com/VoodooI2C/VoodooI2C/pull/514)
>
>1. Always pass buttons as pressure on single-button touchpads (fix https://github.com/VoodooI2C/VoodooI2C/issues/496)
>2. Always disable Force Touch and pass buttons to buttons device on touchpads with two physical buttons (fixes inability to click without finger on the touchpad (https://github.com/VoodooI2C/VoodooI2C/issues/499) and inability to use the right button with default settings)
>3. Disable Force Touch on all touchpads if tap to click is disabled (fixes inability to make a non-force click with default settings, especially in Recovery where there is no System Preferences application)
>
>* NULL -> nullptr, FALSE -> false
>
>* Update VoodooI2CAtmelMXT
>
>* Update VoodooI2CHID
>
>Co-authored-by: Michael Belyaev <usrsse2@me.com>

**Comment by UsedDiscord** [comment]
> Managed to fix the buttons with a reorder of kext plugins and adding voodoops2trackpad + voodooinput in https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/commit/0b6711556207d49135466101fb8e217ab68704dd. I downgraded voodoo kext versions as part of this, but the same fix as proposed above should be applied when updating to newer versions.

I need to try so 🤔... I will try this night (actually 7AM)

If this can be resolve, it's good

**Comment by Qonfused** [comment]
It did work consistently when I tried it then, but there has since been a regression since updating opencore. I believe the PR I just linked addresses this issue.

**Comment by UsedDiscord** [comment]
I know I'm stupid in terms of PR... but how do you get this famous Kext?  and besides, is it normal that the keyboard does not backlight anymore?  I know I'm asking a bit much I think?  😅

**Comment by Qonfused** [comment]
The kext has to be built in xcode (there's some details on that [here](https://voodooi2c.github.io/#Build%20Environment/Building,%20loading,%20and%20installing%20VoodooI2C)) but I'll provide the kext after building + testing it.

**Comment by Qonfused** [comment]
I had already built it in xcode but you can pull built versions from https://dortania.github.io/build-repo/ by commit.

**Comment by Qonfused** [comment]
There are some duplicate entries for voodooinput in Kernel > Add that I'll need to fix.

**Comment by Qonfused** [comment]
Enabling VoodooPS2Trackpad may be a good correction though.

**Comment by Qonfused** [comment]
Changing this broke it and I'm not sure which order + config there is correct :/

**Comment by Qonfused** [comment]
Just fixed kext load order in 6249948b78853d7435534ecf6434b831fec87ac9.

**Comment by UsedDiscord** [comment]
> I know I'm stupid in terms of PR... but how do for get this famous Kext?  and besides, is it normal that the keyboard does not backlight anymore?  I know I'm asking a bit much I think?  😅

Fixed Keyboard does not backlight by reorder the Kexts...

But nothing right now for the Trackpad 😅

**Comment by UsedDiscord** [comment]
> Just fixed kext load order in 6249948b78853d7435534ecf6434b831fec87ac9.

But nothing about Trackpad, it's exact? (See my last comment)

**Comment by UsedDiscord** [comment]
🤔 

Looking at my DSDT and the Device Manager in the ID Compatibles section, there is a difference!

```ASL
Device (GPI0)
        {
            Method (_HID, 0, NotSerialized)  // _HID: Hardware ID
            {
                If ((GPHD == One))
                {
                    Return ("PNP0C02")
                }
[...]
```

However, in the device manager, it is marked PNP0C50

![image](https://user-images.githubusercontent.com/57113628/208305909-e484c163-2b44-471f-9590-414929eefe30.png)

***This could cause a problem, no?***

**Comment by Qonfused** [comment]
It may be a problem if it's occuping a pin that we want for the trackpad or one of the touchscreen displays, but currently they're not GPIO pinned so it shouldn't conflict for now.

I'd note that your firmware can also contain entries for devices that aren't present in your system (even for your specific config of this laptop there could be various models for something like the trackpad, etc).

**Comment by UsedDiscord** [comment]
See : https://github.com/VoodooI2C/VoodooI2C/releases/tag/2.7.1

Will may repair that bug

**Comment by Qonfused** [comment]
It seems the trackpad buttons are using some generic HID interface; they don't use either of the PS2 mouse/trackpad plugins, but VoodooI2CHID. Device manager doesn't show any Synaptics/Elan SMBUS devices either so it's not a regular click pad.

**Comment by UsedDiscord** [comment]
Yes, otherwise we would have seen that it was a SMBUS or Synaptics...

So it would be more a problem related to VoodooI2CHID than VooodooI2C (And we necessarily exclude VoodooPS2 since it only manages the ATK 102 keyboard)

I checked the info.plist of VoodooI2CHID as well, the Compatible ID corresponds well to the one of the periophere manager.
PNC050 (something like that)

**Comment by Qonfused** [comment]
Seems the PR I had linked to before was only fixing physical buttons on an actual trackpad; commit refers to digitiser/transducer buttons:

https://github.com/VoodooI2C/VoodooI2CHID/compare/6fe83a514e90b52bf872ed641627af190ef93011...c1e31809b3be441be0093b0d5aa3aeab1db56080#diff-a17b259a35228d1fa6e960ea2b4eb5931aec9585cee01403b5e89a8da9d54493

**Comment by UsedDiscord** [comment]
That's quite an interesting comment though...

https://github.com/VoodooI2C/VoodooI2C/blob/9ed1dcdb6e375c54d61945fe9415ccb63627d2f2/Multitouch%20Support/Native/VoodooI2CNativeEngine.cpp#L51

Here, we are talking about pressure, which is quite interesting

-----------------

Also here...

https://github.com/VoodooI2C/VoodooI2CHID/blob/640c0c1c7866f130693b92435671eaf07c7af768/VoodooI2CHID/VoodooI2CMultitouchHIDEventDriver.cpp#L170-L179

Here we talk about that: "there can be no secondary button without a primary button".

So, normally, if there is a primary button, you can have a secondary button...

---------------

https://github.com/VoodooI2C/VoodooI2CHID/blob/640c0c1c7866f130693b92435671eaf07c7af768/VoodooI2CHID/VoodooI2CMultitouchHIDEventDriver.cpp#L582-L601

Here we are talking about the Dell latitude 7390 2-1, why?
It's a laptop with 2 buttons like the Asus ZenBook UX481, UX*** ect...
What can interest us is that the left button was set to 2, and the right button was set to 3 and not 1

So, let's study this for the Asus Zenbook...
        

**_ DEFINITION_** : 

<details>
<summary>What is a Transducer?</summary>

(A transducer is an electronic device that converts energy from one form to another, the electrical signal is then turned into a usable direct current or voltage that is easily measured. Some examples of transducers include microphones and loudspeakers.

There are two different types of transducers an input transducer which is a sensor and an output transducer which is an actuator. We offer linear position transducers and pressure transducers here at variohm. Click below to view the different product ranges:)

</details>

**Comment by Qonfused** [comment]
It seems that most other devices with physical trackpad buttons are either connected via SMBus or are Synaptics clickpads; I can't find any other cases where there is an Elan trackpad w/ I2C buttons.

> (uh... Github does not display the code as expected... bruh... Why? Idk because i'have read : [GitHub Help](https://docs.github.com/fr/get-started/writing-on-github/working-with-advanced-formatting/creating-a-permanent-link-to-a-code-snippet)... so... Mystery...)

Apparently, it only embeds the code preview when the permalink is referencing code from the same repository you're commenting on. It's rather problematic on an org level, since you may want to reference code from another project when talking about regressions/dependencies, or just manage a separate issue tracker in another repo.

**Comment by UsedDiscord** [comment]
I try to study a little by reading as I can (I do like that when I did not learn the programming language, I read and try to understand in the native language) [VoodooI2CELAN](https://github.com/VoodooI2C/VoodooI2CELAN).kext to understand how the Kext works...

I noticed that he is looking for things concerning the ASUS firmware... I have to continue to look at the different variants that exist...
```cpp
bool VoodooI2CELANTouchpadDriver::check_ASUS_firmware(UInt8 productId, UInt8 ic_type) {
    if (ic_type == 0x0E) {
        switch (productId) {
            case 0x05 ... 0x07:
            case 0x09:
            case 0x13:
                return true;
        }
    } else if (ic_type == 0x08 && productId == 0x26) {
        return true;
    }
    return false;
}
```
```cpp
 if (check_ASUS_firmware(product_id, ictype)) {
        IOLog("%s::%s ASUS trackpad detected, applying workaround\n", getName(), device_name);
        retVal = write_ELAN_cmd(ETP_I2C_STAND_CMD, ETP_I2C_WAKE_UP);
        if (retVal != kIOReturnSuccess) {
            IOLog("%s::%s Failed to send wake up cmd (workaround)\n", getName(), device_name);
            return false;
        }
```

Besides, we can forget the [Gitter.im](https://gitter.im/alexandred/VoodooI2C/) of Alexandred, it is not attractive in the sense that nobody answers / talks .

So, they don't offer Technical Support anymore, it seems.
The only way to communicate would be Discord? or GitHub of VoodooI2C but in the "Issues" section : [GitHub Issues VoodooI2C](https://github.com/VoodooI2C/VoodooI2C/issues)

> Apparently, it only embeds the code preview when the permalink is referencing code from the same repository you're commenting on. It's rather problematic on an org level, since you may want to reference code from another project when talking about regressions/dependencies, or just manage a separate issue tracker in another repo.

Okay I'see, thanks ! I would use the famous **```** tags to signal code :)

Like that
```cpp
 ``` << This things
```

**Comment by Qonfused** [comment]
@shiecldk Do you encounter the same button mapping issue using this kext load order?
```cpp
VoodooI2C/VoodooI2CServices
VoodooI2C/VoodooGPIO
VoodooI2C/VoodooInput
VoodooI2C
VoodooI2CHID
VoodooPS2Controller
VoodooPS2Controller/VoodooPS2Keyboard
// Omit PS2Mouse / PS2Trackpad plugins
```

For reference, these are the kext versions tested with:
Kext | Version
--- | ---
VoodooI2C | 2.7.1
VoodooI2CHID | 1.0
VoodooPS2 | 2.2.7

**Comment by shiecldk** [comment]
My Ux582 actually doesn't have the physics buttons as do in the 14-inch models. I assume I wouldn't encounter this issue? This is the sequence I use:
<img width="556" alt="Screen Shot 2023-01-29 at 2 04 56 AM" src="https://user-images.githubusercontent.com/42949962/215283522-d9980697-f0be-41a1-9c6a-419d9edf3379.png">

**Comment by UsedDiscord** [comment]
Yeah the UX582 doesn't have physicals buttons, quite strange by the way as Trackpad...: 
![image](https://user-images.githubusercontent.com/57113628/215283671-87f610f2-28b2-4467-b699-66b7a6558362.png)
(Picture from : [Root-Nation](https://root-nation.com/wp-content/uploads/2021/08/ASUS-ZenBook-Pro-Duo-OLED-1.jpg))

**Comment by Qonfused** [comment]
Well, I guess it's a non-issue for you then. I don't know how I haven't noticed this before lol.

**Comment by shiecldk** [comment]
For the temporary fix of the trackpad freezing issue. Do you think it's possible to implement some SSDT codes to turn off the trackpad with FN+F6 (_Q12) like do through sleep and wake to get it back to work?

**Comment by Qonfused** [comment]
My _Q12 method does work for disabling the trackpad; for reference, mine looks like this:
```dsl
Method (_Q12, 0, NotSerialized)  // _Qxx: EC Query, xx=0x00-0xFF
{
    If (!(DSYN & One))
    {
        If (ATKP)
        {
            ^^^^ATKD.IANE (0x6B)
        }
    }
    ElseIf (ATKP)
    {
        ^^^^ATKD.IANE (0x6F)
    }
}
```
I have ATKP enabled in the KBLC SSDT. I should probably move that over to its own SSDT; I didn't realize it was being used elsewhere.

@shiecldk Is the issue here that your trackpad doesn't wake from sleep? I'm not sure if VoodooI2C will reinitialize the device after calling this method, but it should do so anyways after sleep.

**Comment by shiecldk** [comment]
@Qonfused Sorry my texts made you confused lol. I can disable the Trackpad with the _Q12.
I was referring to the random freezing problem on the Trackpad; when the Trackpad is freezed, I can, however, reenable it by putting the laptop to sleep and waking it to get the Trackpad response again. However, it would still freeze sometime at some point.

Hence, I thought maybe we could do some time with the _Q12 method to do the same thing that sleeps and wake the laptop on the Trackpad to temporarily bypass the problem.

**Comment by UsedDiscord** [comment]
I'm thinking of opening an Issue on VoodooI2C's GitHub about the Trackpad buttons swap because I don't feel like I can find a solution...

**Comment by Qonfused** [comment]
> --- @Qonfused https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/issues/8#issuecomment-1354068257 ---
> Managed to fix the buttons with a reorder of kext plugins and adding voodoops2trackpad + voodooinput in [0b67115](https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/commit/0b6711556207d49135466101fb8e217ab68704dd). I downgraded voodoo kext versions as part of this, but the same fix as proposed above should be applied when updating to newer versions.

If this can be reproduced, it should provide a much clearer picture of what is actually going on with the trackpad buttons. This may have happened as a result of a regression in VoodooI2C code.

**Comment by UsedDiscord** [comment]
I will take care of it soon, I am overwhelmed by my law classes... 

I'm sorry... I could be free in 1 or 2 weeks to try to see what it changes

**Comment by Qonfused** [comment]
Looking into this issue again, it seems that versions 2.7+ recognize both trackpad buttons (not just the primary button).

Previous behavior in 2.6.5 not recognizing the second button was first resolved in version 2.7 due to this PR:
- https://github.com/VoodooI2C/VoodooI2C/pull/445

The very next commit in 2.7 pulls in changes from VoodooI2CHID, which notably incorporates some work from 1Revernger1 reworking gvkt's prior PR in [#58](https://github.com/VoodooI2C/VoodooI2CHID/pull/58):
- (all changes) https://github.com/VoodooI2C/VoodooI2CHID/compare/e3328f179c0ac2362fabd45dc42988d9540a9b9f...6fe83a514e90b52bf872ed641627af190ef93011
  - https://github.com/VoodooI2C/VoodooI2CHID/pull/59

^^ This last PR appears directly related to this issue as the trackpad must be using different report ids as a way to distinguish between the physical trackpad device and buttons.

**Comment by Qonfused** [comment]
On Linux, the trackpad uses multiple drivers per the same HID device to handle the touchpad and mouse inputs as separate HID interfaces (as do the touchscreens for mouse/touchscreen/stylus inputs). It does not expose the trackpad and buttons as separate devices.

**Comment by UsedDiscord** [comment]
> Looking into this issue again, it seems that versions 2.7+ recognize both trackpad buttons (not just the primary button).

unless I'm mistaken, the version that was when I created this issue was 2.7... https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/issues/8#issuecomment-1221542362

I'm not sure I understand?

By the way, @gvkt talks about the ZenBook Pro, as well as @1Revenger1, I didn't have the time to look at what they really talk about, but in any case, they talk about the Zenbook...

But since you quoted the issue, you have probably already seen this



> On Linux, the trackpad uses multiple drivers per the same HID device to handle the touchpad and mouse inputs as separate HID interfaces (as do the touchscreens for mouse/touchscreen/stylus inputs). It does not expose the trackpad and buttons as separate devices.

So, Trackpad + Buttons = 1 Device?



----

_(I'm really sorry I can't help much... it's a bit complicated with my courses (university) that take a lot of my time. I try as I can, I doubt I'll be cited in the credits for helping you a little... If I could be cited in the credits as a little help since a little bit the departure of this repo, I would thank you very much)_

**Comment by Qonfused** [comment]
I believe it was also version 2.7 that I used when I (very briefly) reported success. The prior version was 2.6.5 (which is actually the previous release) so I was curious what changes were made that aided in the physical buttons being detected.

Yeah the trackpad itself works as a single device using the same I2C1 data/clock lines. It seems that the physical buttons have a cable that connects to the trackpad itself, which is how it works as a single unit:

Trackpad Buttons | Trackpad Assembly
--- | ---
![image](https://user-images.githubusercontent.com/32466081/221291128-70f9d598-8ac9-46b0-897f-55f63b0e550a.png) | ![image](https://user-images.githubusercontent.com/32466081/221292272-784dc43e-42db-4db8-a909-ccaeebc55926.png)

**Comment by Qonfused** [comment]
> *(I'm really sorry I can't help much... it's a bit complicated with my courses (university) that take a lot of my time. I try as I can, I doubt I'll be cited in the credits for helping you a little... If I could be cited in the credits as a little help since a little bit the departure of this repo, I would thank you very much)*

@UsedDiscord I think this is a fair request. You've been very helpful w/ this project, so I wouldn't undersell the time and effort you've spent here.

**Comment by UsedDiscord** [comment]
Oh?

I discovered that if you press CTRL, the right click works! (both CTRLs work)

So, you have to press CTRL and the right button to have the right click menu working...

But, unfortunately, the left click while also pressing CTRL performs the same action, i.e. opening the right click menu...

So, 

CTRL + C does not work, you have to do Windows + C to get the copy
CTRL + V doesn't work, you have to do Windows + V to get the paste

But

Right click of the trackpad doesn't work, you have to do CTRL + LEFT / RIGHT click to get the right button that works

**Comment by UsedDiscord** [comment]
> @UsedDiscord I think this is a fair request. You've been very helpful w/ this project, so I wouldn't undersell the time and effort you've spent here.

@Qonfused , Thank you so much for your understanding, I appreciate it. This is a project that I like to be involved in, and despite the fact that there was Drama with one person (I won't mention the name, but shiecldk and you Qonfused was also involved) I still want to contribute.

Thanks to you for always supporting this project!

**Comment by Qonfused** [comment]
It seems that ctrl+click is another way to right-click on macOS: https://support.apple.com/guide/mac-help/right-click-mh35853/mac

**Comment by Qonfused** [comment]
I noticed there is a **HIDEventServiceProperties** dictionary under the Precision Touchpad driver (in IOReg), which has a property **MouseButtonMode** set to `OneButton`. There is also a **MouseButtonDivision** key set to `0x37`, which might be used for separating left/right click functions based on x-position?

I currently don't know how these **HIDEventServiceProperties** get injected into the trackpad event driver (most look like macOS injects them). I created an [injector kext](https://github.com/Qonfused/VoodooInjector) for forcing the correct VoodooI2CHID drivers for the touchscreens (what was done previously in #12). If I can figure out how to inject these properties, there may not need to be any modifications made to the trackpad driver to get both physical buttons working.

**Comment by Qonfused** [comment]
I have noticed an interesting quirk where both physical buttons are aligned to the top-left position (0,0). I believe this is due to the one-button mode (or at least the implementation) since this is supposed to be for trackpads that click based on an under-device actuator, which wouldn't need to report a new finger position for the button (as you still have a finger on the trackpad when clicking). Having the physical buttons pressed would only register as a new finger position (at 0,0) when all other fingers are released, which is only problematic when trying to click and drag using one of these buttons.

Here's a video of how this looks like using [FingerMgmt](https://github.com/jnordberg/FingerMgmt):

https://user-images.githubusercontent.com/32466081/221356584-60768769-eb03-41a6-8e24-340926a505f2.mov

**Comment by Qonfused** [comment]
Relevant line:
https://github.com/VoodooI2C/VoodooI2C/blob/17a5f58227a164b426011fd077a5c549766474b3/Multitouch%20Support/Native/VoodooI2CNativeEngine.cpp#L51
> // if it has secondary button, then it will be passed as buttons on the "trackpoint" device

This is what sets the button x,y positions to 0,0:

```cpp
// Commit: https://github.com/VoodooI2C/VoodooI2C/commit/9ab98319b6411c4a4822cc12f840df85cc684bfc
// Mutitouch Support/Native/VoodooI2CNativeEngine.cpp#LN96-103
96  | if (firstTransducer->has_secondary_button) {
97  |     RelativePointerEvent event;
98  |     event.dx = 0;
99  |     event.dy = 0;
100 |     event.buttons = firstTransducer->physical_button.value();
101 |     event.timestamp = timestamp;
102 |     super::messageClient(kIOMessageVoodooTrackpointRelativePointer, voodooInputInstance, &event, sizeof(event));
103 | }
```

**Comment by Qonfused** [comment]
Much older issue (likely pointing to architecutral issues that have since been resolved), but it may also be caused from incorrect scaling:

https://github.com/VoodooI2C/VoodooI2C/issues/108

**Comment by Qonfused** [comment]
> --- https://github.com/acidanthera/VoodooInput/pull/10#issuecomment-602276720 ---
> In VoodooPS2SynapticsTouchpad there are two devices – VoodooInput's MT2 simulator and the old trackpad, and physical buttons (when not remapped into Force Touch) are sent using the old device, so finger touch simulation is not required.

Resurface of this issue: https://github.com/VoodooI2C/VoodooI2C/issues/208

**Comment by Qonfused** [comment]
Currently collecting details for a new VoodooI2C issue (gist): https://gist.github.com/Qonfused/2191a1adf4e46f7cfb61c82938a26361

**Comment by UsedDiscord** [comment]
Thanks for creating it, I had started to look at creating an Issue, but you did it x)

To see the continuation...

**Comment by 1Revenger1** [comment]
https://github.com/torvalds/linux/blob/master/drivers/input/mouse/elan_i2c_core.c#L1005
Seems like a fairly simple change. Just have to read a few more bits from the tp_info details.
https://github.com/VoodooI2C/VoodooI2CELAN/blob/3cb69b61fc4dc4eab127787455d0f1c337a3df0c/VoodooI2CELAN/VoodooI2CELANTouchpadDriver.cpp#L236
Current driver assumes that all trackpads are clickpad, which seems to generally hold for most I2C trackpads.

**Comment by Qonfused** [comment]
This is a newer ELAN trackpad that uses the precision touchpad protocol; I had avoided using the ELAN driver based on that mention here: https://github.com/VoodooI2C/VoodooI2CELAN#supported-touchpads.

The right button works but only as left click after https://github.com/VoodooI2C/VoodooI2C/pull/445.

**Comment by 1Revenger1** [comment]
Ah, I understand. I unfortunately am much less helpful with VI2CHID

**Comment by UsedDiscord** [comment]
First of all, thank you very much for taking the time to write @1Revenger1, it was really a chance to mention you ^^', I didn't think you would answer to tell you everything....

Just a question, do you know someone who would be more qualified on I2CHID? if we can't solve the problem...

**Comment by Qonfused** [comment]
I'll look at refactoring the transducer button reporting, though it might be hard to identify a clean fix for a PR. It's not clear what edge cases are being handled in the current implementation.

**Comment by Qonfused** [comment]
^^ You can subscribe to notifications in the linked issue.

**Comment by UsedDiscord** [comment]
Using the latest version of the Qonfused EFI, I noticed that VoodooInput spams this a lot in kernel-logs:  

**[...]
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
2023-03-26 18:28:19.37 kernel[0]: (VoodooInput) <compose failure [UUID]>
[...]**

also, this regarding AppleSmartBatteryManager, I wonder if there is not a problem with the battery management in macOS? : 

**"2023-03-26 18:28:07.72 kernel[0]: (AppleSmartBatteryManager) Failed to read shutdown data error flags"**

**"2023-03-26 18:28:14.36 kernel[0]: (AppleSmartBatteryManager) Failed to read keyInfo . rc:0x84
2023-03-26 18:28:14.36 kernel[0]: (AppleSmartBatteryManager) SMC Key is not supported."**

**Comment by 1Revenger1** [comment]
The log command can be quite useless. The compose failure is errorings trying to print the logs. I usually just use `sudo dmesg` nowadays.

If the debug version of VoodooInput is being used, it does log each packet being sent.

**Comment by Qonfused** [comment]
The `<compose failure [UUID]>` lines are an issue with Apple's logging (https://github.com/acidanthera/bugtracker/issues/1148). I filter those lines out explicitly in my debug script when pulling logs from syslog; it's a strange issue but they're safe to ignore.

I think those AppleSmartBatteryManager messages are indicating a rtc/shutdown issue. It's probably related to any startup dialogs like 'you shut down your computer because of problem'.

**Comment by UsedDiscord** [comment]
This issue can be marked as Resolved by @1Revenger1 !
 
[VoodooI2CHID.kext.zip](https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/files/11445259/VoodooI2CHID.kext.zip)

Everything is correctly attributed! The left click is the left click, the right click is the right click.

A solved issue! Thank you!

**Comment by UsedDiscord** [comment]
Closed. Free to reopen this issue.

**Comment by Qonfused** [comment]
There's also the https://github.com/VoodooI2C/VoodooI2C/issues/527 issue for tracking. Until a PR is merged, I'll keep this open with the other issue for consistency.

**Comment by Qonfused** [comment]
For now, I'll vendor the built kext with 109fcd5a28f4617da313dc78d0dfaa19d7095d11.

**Comment by Qonfused** [comment]
Fixed in https://github.com/VoodooI2C/VoodooI2CHID/commit/9ebbc8f823e54924dc24a43000a61c3e34cb6800. Thanks @1Revenger1 and @UsedDiscord!
